### PR TITLE
Logging redesign

### LINKED ISSUE #1122: Testability: add log records on test generation phases
**Description**

In order to check phases and their duration we need to have log info records.
Currently we can measure only overall test generation process.
It's needed to analyze which phase is time consuming.

**Expected behavior**

Add info log records for different phases of UnitTestBot test generation.
- Test generation initiated
- Collecting information (classes)
- Fuzzing phase
- Symbolic analysis phase
- Minimization
- Summaries
- Code generation

**Environment**

IntelliJ IDEA (probably add records to idea.log?)

**Potential alternatives**

or separate UnitTestBot log ?

**Context**

_Originally posted by @alisevych in https://github.com/UnitTestBot/UTBotJava/issues/1028#issuecomment-1273121257_

### PR DESCRIPTION
**Pull request by Domonion**
# Description

1. Added Logger model to rd
2. InstrumentedProcess is now sending all logs to EngineProcess, which then writes to its own file
3. Removed instrumentedProcessLogLevel
4. auxiliary log functions
5. stop process rdcall
6. UtRd Loggers redesigned to have the same logging pattern

Fixes #1122

## Type of Change

- Bug fix (non-breaking change which fixes an issue)
- Refactoring (typos and non-functional changes) 
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/tree/CgMethodConstructor.kt [hunk 1]
```diff
@@ -1904,21 +1904,21 @@ open class CgMethodConstructor(val context: CgContext) : CgContextOwner by conte
...
+//                    // if the last statement is a <pre> tag, put the path inside it
+//                    if (lastTag == null || lastTag !is CgDocPreTagStatement) {
+//                        docComment += jvmReportDocumentation
+//                    } else {
+//                        val tagContent = lastTag.content
+//                        docComment.removeLast()
+//                        docComment += CgDocPreTagStatement(tagContent + jvmReportDocumentation)
+//                    }
+//                }
+//            }
```
Comment (sergeypospelov) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/tree/CgMethodConstructor.kt:1921]: Why is it commented?

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/InstrumentedProcessMain.kt [hunk 2]
```diff
@@ -128,7 +145,6 @@ private fun InstrumentedProcessModel.setup(kryoHelper: KryoHelper, watchdog: Idl
         }
     }
     watchdog.wrapActiveCall(invokeMethodCommand) { params ->
-        logger.debug { "received invokeMethod request: ${params.classname}, ${params.signature}" }
```
Comment (sergeypospelov) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/InstrumentedProcessMain.kt:131]: Why did we lost this?

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/ConcreteExecutor.kt [hunk 3]
```diff
@@ -179,12 +185,12 @@ class ConcreteExecutor<TIResult, TInstrumentation : Instrumentation<TIResult>> p
...
-                        proc?.run { process.inputStream.bufferedReader().lines().toList() } ?: emptyList()
-                    } catch (e: Exception) {
-                        emptyList()
-                    }
+//                    instrumentedProcessRunner.errorLogFile,
+//                    try {
+//                        proc?.run { process.inputStream.bufferedReader().lines().toList() } ?: emptyList()
+//                    } catch (e: Exception) {
+//                        emptyList()
+//                    }
```
Comment (sergeypospelov) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/ConcreteExecutor.kt:193]: Remove if we don't need it.

**Review by sergeypospelov — CHANGES_REQUESTED** [review]
Looks OK after fixing remarks, but I didn't dive into details deeply. I guess, you need to update docs.

File: utbot-core/src/main/kotlin/org/utbot/common/Logging.kt [hunk 4]
```diff
@@ -63,20 +63,28 @@ inline fun <T> LoggerWithLogMethod.bracket(
...
     return try {
         block()
     } catch (e: Throwable) {
-        this.error("Exception occurred", e)
+        this.error(message, e)
         throw e
     }
+}
+
+inline fun <T> silent(block: () -> T): T? {
```
[the remark this answers is not in the document]
Comment (Domonion) [reply utbot-core/src/main/kotlin/org/utbot/common/Logging.kt:84]: Used - ProcessWithRdServer

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/InstrumentedProcessMain.kt (hunk 2 above)
Comment (Domonion) [reply utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/process/InstrumentedProcessMain.kt:131]: fixed

**Review by sergeypospelov — APPROVED** [review]

**Comment by alisevych** [comment]
@Domonion Please add logging of Summarization phase: started - ended.

File: utbot-rd/src/main/kotlin/org/utbot/rd/loggers/UtRdRemoteLogger.kt [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        val logLevelValues = LogLevel.values()
+        private val threadLocalExecutingBackingFiled: ThreadLocal<CountingSet<UtRdRemoteLogger>> =
+            threadLocalWithInitial { CountingSet() }
+
+        // !!! IMPORTANT !!! Don't use 'by ThreadLocal' to avoid slow reflection initialization
+        internal val threadLocalExecuting get() = threadLocalExecutingBackingFiled.get()
+    }
+
+    override fun isEnabled(level: LogLevel): Boolean {
+        // On every protocol send/receive event RD to its own loggers.
```
Comment (EgorkaKulikov) [inline utbot-rd/src/main/kotlin/org/utbot/rd/loggers/UtRdRemoteLogger.kt:28]: sends

File: utbot-rd/src/main/kotlin/org/utbot/rd/loggers/UtRdRemoteLogger.kt [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+) : Logger {
+    private val logLevel: LogLevel = logLevelValues[loggerModel.getCategoryMinimalLogLevel.startBlocking(category)]
+
+    companion object {
+        val logLevelValues = LogLevel.values()
+        private val threadLocalExecutingBackingFiled: ThreadLocal<CountingSet<UtRdRemoteLogger>> =
+            threadLocalWithInitial { CountingSet() }
+
+        // !!! IMPORTANT !!! Don't use 'by ThreadLocal' to avoid slow reflection initialization
+        internal val threadLocalExecuting get() = threadLocalExecutingBackingFiled.get()
```
Comment (EgorkaKulikov) [inline utbot-rd/src/main/kotlin/org/utbot/rd/loggers/UtRdRemoteLogger.kt:24]: do not change it to by Threadlocal

File: utbot-summary/src/main/kotlin/org/utbot/summary/Summarization.kt [hunk 7]
```diff
@@ -36,7 +38,17 @@ import soot.SootMethod
...
+    "-------------------------------------------------------------------------------------------------\n" +
+    "------------------------Summarization started for ${this.size} test cases------------------------\n" +
+    "-------------------------------------------------------------------------------------------------"
+    }) {
+    this.map {
+        it.summarizeOne(searchDirectory, sourceFile)
+    }
+}
+
+fun UtMethodTestSet.summarizeOne(searchDirectory: Path, sourceFile: File?): UtMethodTestSet = logger.info().bracket({ "Summarization for ${this.method}"} ){
```
Comment (EgorkaKulikov) [inline utbot-summary/src/main/kotlin/org/utbot/summary/Summarization.kt:51]: Make it private please

**Review by EgorkaKulikov — COMMENTED** [review]
LVGNP2m

File: docs/InterprocessLogging.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+This will reduce number of IO operations and use log4j2 buffer more efficiently. The cost it that 
+when process terminates - log4j2 terminates logging service before buffer is flushed and 
+you will lose last portion of logs. This might be undesired behaviour in tests and debugging, 
+but probably acceptable in release.
+
+### Docker and Gradle
+
+To see logs in Gradle from console, Docker and CI - add following `build.gradle.kts`:
+```kotlin
```
[the remark this answers is not in the document]
Comment (Domonion) [reply docs/InterprocessLogging.md:157]: It is already in javadoc, it is just mentioned for clarification that this option is available

**Review by EgorkaKulikov — APPROVED** [review]

**Pull request merged**